File: lib/server/route-analysis.ts
import type {
  ExposureLevel,
  PollenSignal,
  ResolvedWaypoint,
  RouteAnalysisRequest,
  RouteAnalysisResponse,
  RouteCandidate,
  WeatherSignal,
} from "@/lib/types";
import { scoreRoutes } from "@/lib/scoring";
import { midpoint } from "@/lib/utils";
import {
  buildFallbackRoutes,
  computeAlternativeWalkingRoutes,
  geocodeAddress,
} from "@/lib/server/google-maps";
import { generateGroundedCopy } from "@/lib/server/gemini";
import { getPollenSignal } from "@/lib/server/google-pollen";
import { getWeatherSignal } from "@/lib/server/google-weather";

export interface RouteAnalysisDependencies {
  geocode: (address: string) => Promise<ResolvedWaypoint>;
  routes: typeof computeAlternativeWalkingRoutes;
  pollen: typeof getPollenSignal;
  weather: typeof getWeatherSignal;
  copy: typeof generateGroundedCopy;
}

const DEFAULT_WEATHER: WeatherSignal = {
  description: "Weather fallback active; using calm spring conditions.",
  windSpeedMph: 8,
  humidity: 54,
  temperatureF: 61,
};

const DEFAULT_POLLEN: PollenSignal = {
  treeIndex: 3,
  grassIndex: 1,
  weedIndex: 1,
  summary: "Live pollen unavailable; using tree-grid-weighted fallback.",
};

export const defaultDependencies: RouteAnalysisDependencies = {
  geocode: geocodeAddress,
  routes: computeAlternativeWalkingRoutes,
  pollen: getPollenSignal,
  weather: getWeatherSignal,
  copy: generateGroundedCopy,
};

export async function analyzeRouteRequest(
  request: RouteAnalysisRequest,
  dependencies: RouteAnalysisDependencies = defaultDependencies,
): Promise<RouteAnalysisResponse> {
  validateRequest(request);

  const origin = await resolveWaypoint(request.origin, dependencies);
  const destination = await resolveWaypoint(request.destination, dependencies);
  const centerPoint = midpoint([origin.location, destination.location]);
  const fallbackMode: string[] = [];

  const routes = await getRoutesWithFallback(origin, destination, dependencies, fallbackMode);
  const weather = await getWeatherWithFallback(centerPoint, dependencies, fallbackMode);
  const pollen = await getPollenWithFallback(centerPoint, dependencies, fallbackMode);
  const scoredRoutes = scoreRoutes(routes, request.profile, weather, pollen);
  const routingMode = request.profile.knowsTreeTriggers && request.profile.triggers.length
    ? "specific-tree-triggers"
    : "general-tree-avoidance";

  const topArea = scoredRoutes[0]?.dominantArea ?? "Central Manhattan";
  const topLevel = scoredRoutes[0]?.dominantLevel ?? "moderate";
  const copy = await dependencies.copy({
    profile: request.profile,
    routes: scoredRoutes.map((entry) => entry.candidate),
    weather,
    pollen,
    areaName: topArea,
    burdenLevel: topLevel,
    routingMode,
  });

  const enrichedRoutes = scoredRoutes.map((entry) => applyGeneratedCopy(entry.candidate, copy.routeExplanations));

  return {
    originResolved: origin.address,
    destinationResolved: destination.address,
    originPoint: origin.location,
    destinationPoint: destination.location,
    summary: copy.summary,
    routingMode,
    dataSources: [
      "Google Maps Routes API",
      "Google Maps JavaScript API",
      "Gemini via Google GenAI SDK",
      "NYC 2015 Street Tree Census",
    ],
    routes: enrichedRoutes,
    civicInsight: {
      areaName: topArea,
      treeBurdenLevel: topLevel,
      summary: copy.civicSummary,
    },
    weather,
    pollen,
    fallbackMode,
  };
}

async function resolveWaypoint(
  waypoint: RouteAnalysisRequest["origin"],
  dependencies: RouteAnalysisDependencies,
): Promise<ResolvedWaypoint> {
  if (waypoint.location) {
    return {
      address: waypoint.address || "Selected point",
      location: waypoint.location,
    };
  }

  return dependencies.geocode(waypoint.address);
}

async function getRoutesWithFallback(
  origin: ResolvedWaypoint,
  destination: ResolvedWaypoint,
  dependencies: RouteAnalysisDependencies,
  fallbackMode: string[],
) {
  try {
    const liveRoutes = await dependencies.routes(origin.location, destination.location);
    if (liveRoutes.length) {
      return liveRoutes;
    }
  } catch {
    fallbackMode.push("fallback-routes");
  }

  return buildFallbackRoutes(origin.location, destination.location);
}

async function getWeatherWithFallback(
  point: ResolvedWaypoint["location"],
  dependencies: RouteAnalysisDependencies,
  fallbackMode: string[],
) {
  try {
    return await dependencies.weather(point);
  } catch {
    fallbackMode.push("fallback-weather");
    return DEFAULT_WEATHER;
  }
}

async function getPollenWithFallback(
  point: ResolvedWaypoint["location"],
  dependencies: RouteAnalysisDependencies,
  fallbackMode: string[],
) {
  try {
    return await dependencies.pollen(point);
  } catch {
    fallbackMode.push("fallback-pollen");
    return DEFAULT_POLLEN;
  }
}

function applyGeneratedCopy(
  candidate: RouteCandidate,
  generated: Record<string, { explanation: string; rationale: string[] }>,
): RouteCandidate {
  const copy = generated[candidate.id];
  if (!copy) {
    return candidate;
  }

  return {
    ...candidate,
    explanation: copy.explanation,
    rationale: copy.rationale?.length ? copy.rationale : candidate.rationale,
  };
}

function validateRequest(request: RouteAnalysisRequest) {
  if (!request.origin?.address && !request.origin?.location) {
    throw new Error("Origin is required.");
  }

  if (!request.destination?.address && !request.destination?.location) {
    throw new Error("Destination is required.");
  }

  if (request.profile.knowsTreeTriggers && !request.profile?.triggers?.length) {
    throw new Error("Choose at least one tree trigger or switch to general tree avoidance.");
  }
}

File: backend/app/route_analysis.py
from __future__ import annotations

from .geometry import midpoint
from .models import CivicInsight, ResolvedWaypoint, RouteAnalysisRequest, RouteAnalysisResponse, RouteCandidate, RoutingMode
from .providers import (
    DEFAULT_POLLEN,
    DEFAULT_WEATHER,
    build_fallback_routes,
    compute_alternative_walking_routes,
    generate_grounded_copy,
    geocode_address,
    get_pollen_signal,
    get_weather_signal,
)
from .scoring import score_routes


async def analyze_route_request(request: RouteAnalysisRequest) -> RouteAnalysisResponse:
    validate_request(request)

    origin = await resolve_waypoint(request.origin)
    destination = await resolve_waypoint(request.destination)
    center_point = midpoint([origin.location, destination.location])
    fallback_mode: list[str] = []

    routes = await get_routes_with_fallback(origin.location, destination.location, fallback_mode)
    weather = await get_weather_with_fallback(center_point, fallback_mode)
    pollen = await get_pollen_with_fallback(center_point, fallback_mode)
    scored_routes = score_routes(routes, request.profile, weather, pollen)
    routing_mode: RoutingMode = (
        "specific-tree-triggers"
        if request.profile.knowsTreeTriggers and request.profile.triggers
        else "general-tree-avoidance"
    )

    top_area = scored_routes[0]["dominant_area"] if scored_routes else "Central Manhattan"
    top_level = scored_routes[0]["dominant_level"] if scored_routes else "moderate"
    copy = await generate_grounded_copy(
        profile=request.profile,
        routes=[entry["candidate"] for entry in scored_routes],
        weather=weather,
        pollen=pollen,
        area_name=top_area,
        burden_level=top_level,
        routing_mode=routing_mode,
    )

    enriched_routes = [
        apply_generated_copy(entry["candidate"], copy["routeExplanations"])
        for entry in scored_routes
    ]

    return RouteAnalysisResponse(
        originResolved=origin.address,
        destinationResolved=destination.address,
        originPoint=origin.location,
        destinationPoint=destination.location,
        summary=copy["summary"],
        routingMode=routing_mode,
        dataSources=[
            "Google Maps Geocoding API",
            "Google Routes API",
            "Google Pollen API",
            "Google Weather API",
            "Gemini via Google GenAI SDK",
            "NYC 2015 Street Tree Census",
        ],
        routes=enriched_routes,
        civicInsight=CivicInsight(
            areaName=top_area,
            treeBurdenLevel=top_level,
            summary=copy["civicSummary"],
        ),
        weather=weather,
        pollen=pollen,
        fallbackMode=fallback_mode,
    )


def validate_request(request: RouteAnalysisRequest):
    if not request.origin.address and not request.origin.location:
        raise ValueError("Origin is required.")

    if not request.destination.address and not request.destination.location:
        raise ValueError("Destination is required.")

    if request.profile.knowsTreeTriggers and not request.profile.triggers:
        raise ValueError("Choose at least one tree trigger or switch to general tree avoidance.")


async def resolve_waypoint(waypoint):
    if waypoint.location:
        return ResolvedWaypoint(address=waypoint.address or "Selected point", location=waypoint.location)

    return await geocode_address(waypoint.address)


async def get_routes_with_fallback(origin, destination, fallback_mode: list[str]):
    try:
        live_routes = await compute_alternative_walking_routes(origin, destination)
        if live_routes:
            return live_routes
    except Exception:
        fallback_mode.append("fallback-routes")

    return build_fallback_routes(origin, destination)


async def get_weather_with_fallback(point, fallback_mode: list[str]):
    try:
        return await get_weather_signal(point)
    except Exception:
        fallback_mode.append("fallback-weather")
        return DEFAULT_WEATHER


async def get_pollen_with_fallback(point, fallback_mode: list[str]):
    try:
        return await get_pollen_signal(point)
    except Exception:
        fallback_mode.append("fallback-pollen")
        return DEFAULT_POLLEN


def apply_generated_copy(candidate: RouteCandidate, generated: dict[str, dict]):
    if not isinstance(generated, dict):
        return candidate

    copy = generated.get(candidate.id)
    if not isinstance(copy, dict):
        return candidate

    rationale = copy.get("rationale") if isinstance(copy.get("rationale"), list) else candidate.rationale
    explanation = copy.get("explanation") if isinstance(copy.get("explanation"), str) else candidate.explanation

    return RouteCandidate(
        **{
            **candidate.model_dump(),
            "explanation": explanation,
            "rationale": rationale,
        }
    )
